### PR TITLE
Expose number of subsamples in ImageItem auto-level determination

### LINKED ISSUE #1608: setImage(..) autoLevels not considering all data when calculating min,max
### Short description

Since `setImage` skips pixels for large images, the max and min values are off. Therefore the `levels` determined from the data is incorrect, causing an apparent empty image.

Please see `setImage` in `ImageItem.py` and the screenshot below.

image

### Expected behavior

The levels should have been determined correctly.

### Real behavior

Since only a subset of the image is considered, the min and max values are off resulting in a wrong look up table. Note that in my field (physics) single pixel 'hits' are very common, and this behavior results in empty images _sometimes_, depending on whether the pixel happens to be part of the image considered for the minimum and maximum values. I spent multiple hours debugging this..

In any case, thanks for all the work you all have put into this project.

### PR DESCRIPTION
**Pull request by NilsNemitz**
#1608 reports that `ImageItem.setImage()` only looks at a subset of pixels to determine the automatic level range.

While that is certainly an efficient approach for normal images, it is not uncommon in scientific data acquisition to look for the presence or absence of events in a very small number of pixels. In that case, it is necessary to increase the number of inspected pixels, or even check for minimum / maximum across the entire image. Right now, the maximum number of samples is hardcoded to 65536, not a large number when megapixel sensors are easily available.

This PR exposes the target number of samples in the `setImage()` call as a new 'levelSamples' keyword argument.
By setting this to the full number of pixels, levels will consider the entire image. Otherwise the image is subsampled by a factor of two in each direction until the remaining pixels are no longer more than the requested levelSamples. This is the same behavior as before, just with a variable endpoint.

closes #1608

### CODE REVIEW

**Comment by j9ac9k** [comment]
@ixjlyons  you have input on this?  This looks entirely reasonable to me.

File: pyqtgraph/graphicsItems/ImageItem.py [hunk 1]
```diff
@@ -299,8 +302,11 @@ def setImage(self, image=None, autoLevels=None, **kargs):
                 autoLevels = True
         if autoLevels:
             img = self.image
-            while img.size > 2**16:
+            level_samples = kargs.pop('levelSamples', 2**16) # use specified value or default of 2**16, remove from kargs
+            if level_samples < 2: level_samples = 2 # avoid endless loop, and keep at least two values for min, max
+            while img.size > level_samples:
                 img = img[::2, ::2]
+            print( img.size )
```
Comment (j9ac9k) [inline pyqtgraph/graphicsItems/ImageItem.py:309]: We probably won't want this print statement in the merged code-base.

Comment (j9ac9k) [reply pyqtgraph/graphicsItems/ImageItem.py]: resolved at @NilsNemitz 's request

Comment (NilsNemitz) [reply pyqtgraph/graphicsItems/ImageItem.py]: Ooops, my bad. Thank you for taking that out!

File: pyqtgraph/graphicsItems/ImageItem.py [hunk 2]
```diff
@@ -242,6 +242,9 @@ def setImage(self, image=None, autoLevels=None, **kargs):
         levels             (min, max) The minimum and maximum values to use when rescaling the image
                            data. By default, this will be set to the minimum and maximum values
                            in the image. If the image array has dtype uint8, no rescaling is necessary.
+        levelSamples       (default: 65536) When determining minimum and maximum values, ImageItem
+                           only inspects a subset of pixels no larger than this number.
```
[the remark this answers is not in the document]
Comment (NilsNemitz) [reply pyqtgraph/graphicsItems/ImageItem.py:246]: Very good point.
I did not think about the 3d stacks of images also supported by this. I also think that the number should be per image slice.

Could you specify what the default behavior should be?
Right now it samples 65536 points across the entire set, and not even randomly, but in the same grid for each slice. I cannot see that as ideal for a deep stack of images. I am also pretty sure that a stack more than 65536 deep will hang the code now.
So I am in favor of changing the default behavior (for stacked images only).
Is 65536 samples per slice(=single image) okay?


Please explain why "grayscale" comes up here.... Does this even run for multichannel images? I thought the levels are specifically used in color-mapped images.

[the remark this answers is not in the document]
Comment (ixjlyons) [reply pyqtgraph/graphicsItems/ImageItem.py:246]: As far as I know, `ImageItem` is only meant to work with a single image, so it's either a 2D array with shape `(px, py)` or 3D with shape `(px, py, ch)` where `ch` can be 3 for RGB or 4 for RGBA. Maybe you're thinking of `ImageView`, which does take stacks of images.

Yeah "grayscale" wasn't precise, I was just referring to monochrome as opposed to RGB(A).

[the remark this answers is not in the document]
Comment (NilsNemitz) [reply pyqtgraph/graphicsItems/ImageItem.py:246]: I really like `while img.shape[0]*img.shape[1] > level_samples` as the loop condition. It requires no additional checks for array dimensions, and no separation of code paths for the 2D arrays of a single channel image and the 3D arrays of RGB / RGBA images.

This *does* change the default behaviour now to perform the min/max search over 65536 * 3 values for RGB images, and 65536 * 4 values for RGBA. But the previous behaviour of sampling less pixels after adding an alpha channel... that doesn't make much sense to me.

This will have some impact on performance, but the user *can* now make it run even faster than before if they are willing to reduce the levelSamples further.

What do you think, @ixjlyons ?

[the remark this answers is not in the document]
Comment (NilsNemitz) [reply pyqtgraph/graphicsItems/ImageItem.py:246]: Sorry for the confusion of 2D/3D images. All my applications for this are single channel :)
Thank you for setting this straight.

**Comment by NilsNemitz** [comment]
Code is updated to always consider pixels now (regardless of these having 1, 3 or 4 channels).
I changed some comments and set a fixed low sample limit, since we look at *pixels*.
The low limit is 4, since we want to keep a 2x2 image rather than downsampling it to 1x1.

**Review by ixjlyons — APPROVED** [review]
I think I'm ok with the slight change in behavior. Others are welcome to chime in.

For clarity since the comment chains on PRs can be kinda hard to follow: `setImage` would previously downsample the input image until its *overall* size including color channels was <= 2^16. With this PR it does so until the image width*height <= 2^16 by default. Default behavior for monochrome (single-channel) images is unaffected.

**Comment by j9ac9k** [comment]
@outofculture @pijyoi you two I know also use ImageItem's a decent amount, you two have anything to add?

**Comment by pijyoi** [comment]
There is an assumption (also in the original code) that decimation simultaneously in rows and cols is the right thing to do. For a skinny array, it might be desirable to decimate along the longer dimension first.

**Comment by pijyoi** [comment]
Wouldn't a solution for #1608 be for the user to compute the levels manually and pass them in as a parameter, thereby disabling autoLevels?

The issue seems to be that the user was not aware that autoLevels would use a decimated array. The changes here still require the user to be aware that decimation takes place by default.

**Comment by NilsNemitz** [comment]
I agree that #1608 is primarily a user-awareness issue, with the description of `levels` giving no hint that the automatic method is not actually looking at all pixels. We are now adding a hint in the description of the new levelSamples parameter:
```
levelSamples  (default: 65536) When determining minimum and maximum values, ImageItem
              only inspects a subset of pixels no larger than this number.
```
But it might be a good idea to extend the descriptions of levels, too. How about:
```
levels      (min, max) The minimum and maximum values to use when rescaling the image
            data. By default, this will be set to the minimum and maximum values
            obtained from sampling the image according to levelSamples. 
            If the image array has dtype uint8, no rescaling is necessary.
```
@pijyoi , do you have an idea for an elegant solution to decimate towards a more square array? That would be nice to add.
On the other hand, if I am effectively running a line camera by looking at a long narrow strip of pixels, then I might be expecting to see detail primarily along this long axis. Not sure if decimating this axis first is the best thing to do, then.

In general, I would think the assumption of a square-ish image is more reasonable than the assumption that any image can be auto-leveled with the same number of checks...

**Comment by NilsNemitz** [comment]
(Added)

**Comment by pijyoi** [comment]
>do you have an idea for an elegant solution to decimate towards a more square array?

```python
import numpy as np 
arr = np.zeros((31, 12345, 3), dtype=np.uint8)

while True:
    print(arr.shape)
    h, w = arr.shape[:2]
    if h * w < 65536:
        break
    if h > w:
        arr = arr[::2, ::]
    else:
        arr = arr[::, ::2]
```

**Comment by NilsNemitz** [comment]
Hmm. Stupid question: 
Why are we iteratively down-sampling?
Can't we just calculate a useful stride right-away?

**Comment by pijyoi** [comment]
There seems to be an existing method ImageItem::quickMinMax() that already does the same decimate-largest-axis logic.

**Comment by NilsNemitz** [comment]
Dear @pijyoi , sorry, it took me a while to get around to looking at this again.
I changed the code to use the `QuickMinMax()` method in the initial level determination.
It turned out that this method now throws a `FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated`.

I swapped the old code for the functionally identical code based on your March 20 comment.

**Comment by pijyoi** [comment]
I think using ```data = data[tuple(sl)]``` in the original quickMinMax should clear the warning.

**Comment by NilsNemitz** [comment]
Thank you @pijyoi , but actually I like your new method (from March 20) better:
The previous code
```python
while data.size > targetSize:
    ax = self._xp.argmax(data.shape)
    sl = [slice(None)] * data.ndim
    sl[ax] = slice(None, None, 2)
    data = data[sl]
```
tries to be short and general at the cost of readability. Adding a a cast to tuple will also not make it faster.

Should we keep the March 20 code
``` python
while True:
    h, w = data.shape[:2]
    if h * w <= targetSize: break
    if h > w:
        data = data[::2, ::] # downsample first axis
    else:
        data = data[::, ::2] # downsample second axis
```
as a permanent replacement?

**Comment by pijyoi** [comment]
There's a change in behavior to quickMinMax(). Old behavior uses ```data.size```, i.e. all elements across all dimensions. The new behavior only considers the first 2 dimensions. Though ImageItem::quickMinMax() doesn't even seem to be used by the library at all. So it's a question of whether you want to preserve behavior for some hypothetical end-user that may or may not exist...

ImageView::quickMinMax() also exists (with a different API) and is actually used internally.

**Comment by NilsNemitz** [comment]
@pijyoi , nicely spotted.
I wonder if the difference in downsampling factors for color vs. monochrome images is actually intended.
It does look like an oversight to me, making the same assumption that h * w = size.

So I would still propose to change the quickMinMax function in ImageItem.

We can add a warning to the documentation if that is required. But I suspect that the min-max calculations have the most applications where data is monochrome... and in that case, the implementations should behave effectively identical.

**Comment by j9ac9k** [comment]
LGTM, thanks for the update @NilsNemitz

**Comment by leon-vv** [comment]
I was the user who was unaware of the min/max behavior and opened #1608. I just wanted to thank everyone for the substantial time and effort invested to fix the issue. Really shows the quality of the project

**Comment by j9ac9k** [comment]
@leon-vv thanks for the kind words, we try our best!  Also a lot of us maintainers are unaware of behavior in the library so issues like the one you opened can be either indicate bugs or areas within documentation we should be doing better; and can be an exercise in us trying to improve as well.
